File: src/main/scala/com/leetcode/easy/RomanToInteger.scala
package com.leetcode.easy

object RomanToInteger {
  def romanToInt(s: String): Int = {
    calculate(s.toCharArray.toList, 0)
  }
  
  def calculate(list: List[Char], n: Int): Int = {
    list match {
      case Nil => n
      case 'I' :: 'V' :: t => calculate(t, n + 4)
      case 'I' :: 'X' :: t => calculate(t, n + 9)
      case 'I' :: t => calculate(t, n + 1)
      case 'V' :: t => calculate(t, n + 5)
      case 'X' :: 'L' :: t => calculate(t, n + 40)
      case 'X' :: 'C' :: t => calculate(t, n + 90)
      case 'L' :: t => calculate(t, n + 50)
      case 'X' :: t => calculate(t, n + 10)
      case 'C' :: 'D' :: t => calculate(t, n + 400)
      case 'C' :: 'M' :: t => calculate(t, n + 900)
      case 'C' :: t => calculate(t, n + 100)
      case 'D' :: t => calculate(t, n + 500)
      case 'M' :: t => calculate(t, n + 1000)
    }
  }
  
  def main(args: Array[String]): Unit = {
    println(romanToInt("III"))
    println(romanToInt("IV"))
    println(romanToInt("IX"))
    println(romanToInt("LVIII"))
    println(romanToInt("MCMXCIV"))
  }
}

File: src/main/java/com/leecode/easy/RomanToInteger.java
package com.leecode.easy;

public class RomanToInteger {
    
    public int romanToInt(String s) {
        char[] chars = s.toCharArray();
        int sum = 0;
        for (int i = 0; i < chars.length; ++i) {
            char current = chars[i];
            int currentValue = getValue(current);
            boolean hasNext = i < chars.length - 1;
            int item = currentValue;
            if (hasNext) {
                char next = chars[i + 1];
                int nextValue = getValue(next);
                if (currentValue < nextValue) {
                    item = nextValue - currentValue;
                    ++i;
                }
            }
            sum += item;
        }
        return sum;
    }
    
    private int getValue(char c) {
        switch (c) {
            case 'I':
                return 1;
            case 'V':
                return 5;
            case 'X':
                return 10;
            case 'L':
                return 50;
            case 'C':
                return 100;
            case 'D':
                return 500;
            case 'M':
                return 1000;
        }
        return 0;
    }
    
    public static void main(String[] args) {
        RomanToInteger app = new RomanToInteger();
        System.out.println(app.romanToInt("III"));
        System.out.println(app.romanToInt("IV"));
        System.out.println(app.romanToInt("IX"));
        System.out.println(app.romanToInt("LVIII"));
    }
}